### PR TITLE
Add expression aggregator

### PR DESCRIPTION
**Pull request by mzganec**
This version sets the JSON field `exp_matrix` but ignores expression matrix on re-runs.

### CODE REVIEW

File: resolwe_bio/tools/expression_aggregator.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    """Get box plot statistic for expressions of a single gene and attribute."""
+    min_val = min(expression)
+    max_val = max(expression)
+    median = np.percentile(expression, 50.0)
+    q1 = np.percentile(expression, 25.0)
+    q3 = np.percentile(expression, 75.0)
+    iqr = q3 - q1
+    lowerwhisker = max(min_val, q1 - 1.5 * iqr)
+    upperwhisker = min(max_val, q3 + 1.5 * iqr)
+    return {'attribute': attribute,
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Add a new line after `{` and indent keys only for 4 spaces.

File: resolwe_bio/tools/expression_aggregator.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    """Return expressions of a gene with matching descriptor."""
+    indices = get_indices(descriptors, descriptor)
+    return [expression[gene] for i, expression in enumerate(expressions) if i in indices and gene in expression]
+
+
+def get_expressions(fnames=[], sep='\t', descriptors=[]):
+    """Read expressions from files."""
+    expressions = [get_expression(fname, sep) for fname in fnames]
+    genes = set().union(*[expression.keys() for expression in expressions])
+    return {gene: {descriptor: get_values(expressions, descriptors, gene, descriptor)
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: return {
        gene: {
            descriptor: get_values(expressions, descriptors, gene, descriptor)
            for descriptor in set(descriptors)
            if get_values(expressions, descriptors, gene, descriptor)
        } 
        for gene in genes
    }

File: resolwe_bio/tools/expression_aggregator.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+def get_values(expressions, descriptors, gene, descriptor):
+    """Return expressions of a gene with matching descriptor."""
+    indices = get_indices(descriptors, descriptor)
+    return [expression[gene] for i, expression in enumerate(expressions) if i in indices and gene in expression]
+
+
+def get_expressions(fnames=[], sep='\t', descriptors=[]):
+    """Read expressions from files."""
+    expressions = [get_expression(fname, sep) for fname in fnames]
+    genes = set().union(*[expression.keys() for expression in expressions])
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Can you omit asterisk and parentheses?

    genes = set().union(expression.keys() for expression in expressions)

File: resolwe_bio/tools/expression_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    return {'attribute': attribute,
+            'gene': gene,
+            'exp_types': exp_types,
+            'min': min_val,
+            'max': max_val,
+            'median': median,
+            'q1': q1,
+            'q3': q3,
+            'lowerwhisker': lowerwhisker,
+            'upperwhisker': upperwhisker}
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: New line before `}`.

File: resolwe_bio/tools/expression_aggregator.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+def main():
+    """Compute expression statistics."""
+    args = get_args()
+    expressions = get_expressions(fnames=args.expressions, descriptors=args.descriptors)
+    statistics = get_statistics(expressions, args.expression_types)
+    output_json(statistics, args.output_box_plot)
+    output_json(expressions, args.output_expressions)
+
+
+main()
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: if __name__ == "__main__":
        main()

File: resolwe_bio/tests/processes/test_expression_aggregator.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        sample3 = Sample.objects.get(data=expression3)
+        sample3.descriptor['sample'] = {'organism': 'Mus musculus'}
+        sample3.save()
+
+        expression4 = self.prepare_expression(f_rc='exp_2_rc.tab.gz', f_exp='exp_2_tpm.tab.gz', f_type='TPM')
+        sample4 = Sample.objects.get(data=expression4)
+        sample4.descriptor['sample'] = {'organism': 'Mus musculus'}
+        sample4.save()
+
+        inputs = {'exps': [expression1.id, expression2.id, expression3.id, expression4.id],
```
Comment (dblenkus) [inline resolwe_bio/tests/processes/test_expression_aggregator.py]: New lines after `{` and before `}`.

File: resolwe_bio/tests/processes/test_expression_aggregator.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                  'group_by': 'sample.organism'}
+        expression_aggregator = self.run_process('expression-aggregator', inputs)
+        for data in Data.objects.all():
+            self.assertStatus(data, Data.STATUS_DONE)
```
Comment (dblenkus) [inline resolwe_bio/tests/processes/test_expression_aggregator.py]: `run_process` already checks that status of resulting object is set to `DONE`.

File: resolwe_bio/tests/processes/test_expression_aggregator.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+        self.assertFile(expression_aggregator, 'exp_matrix', 'exp_matrix.json')
+        self.assertJSON(expression_aggregator, expression_aggregator.output['box_plot'],
```
Comment (dblenkus) [inline resolwe_bio/tests/processes/test_expression_aggregator.py]: self.assertJSON(
        expression_aggregator, expression_aggregator.output['box_plot'], '', 'box_plot.json.gz'
    )

File: resolwe_bio/tests/processes/test_expression_aggregator.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# pylint: disable=missing-docstring
+from resolwe.flow.models import Data
+from resolwe_bio.models import Sample
+from resolwe_bio.utils.test import BioProcessTestCase
+
+
+class ExpressionAggregatorTestCase(BioProcessTestCase):
+    def test_expression_aggregator(self):
+        expression1 = self.prepare_expression(f_rc='exp_1_rc.tab.gz', f_exp='exp_1_tpm.tab.gz', f_type='TPM')
```
Comment (dblenkus) [inline resolwe_bio/tests/processes/test_expression_aggregator.py]: Use helper function for creating samples.

File: resolwe_bio/tools/expression_aggregator.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    parser.add_argument('-e', '--expressions', nargs='+', help='Expressions', required=True)
+    parser.add_argument('-d', '--descriptors', nargs='+', help='Descriptors', required=True)
+    parser.add_argument('-t', '--expression-types', nargs='+', help='Expression types')
+    parser.add_argument('-a', '--aggregator', help='Aggregator', required=False)
+    parser.add_argument('-b', '--output_box_plot', help='Output box plot', required=False)
+    parser.add_argument('-x', '--output_expressions', help='Output expressions', required=False)
+    return parser.parse_args()
+
+
+def get_expression(fname=None, sep='\t'):
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: I think `load_expressions` would be more suitable name.

File: resolwe_bio/tools/expression_aggregator.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+def get_expressions(fnames=[], sep='\t', descriptors=[]):
+    """Read expressions from files."""
+    expressions = [get_expression(fname, sep) for fname in fnames]
+    genes = set().union(*[expression.keys() for expression in expressions])
+    return {gene: {descriptor: get_values(expressions, descriptors, gene, descriptor)
+                   for descriptor in set(descriptors)
+                   if get_values(expressions, descriptors, gene, descriptor)} for gene in genes}
+
+
+def get_statistic(expression, gene, attribute, exp_types):
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: I don't like having multiple functions with almost the same name. Maybe rename it to `generate_statistic`?

File: resolwe_bio/tools/expression_aggregator.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python2
```
Comment (mstajdohar) [inline resolwe_bio/tools/expression_aggregator.py]: All new scripts should be in Python 3, unless there is a good reason not to.

File: resolwe_bio/tools/expression_aggregator.py (hunk 3 above)
Comment (mzganec) [reply resolwe_bio/tools/expression_aggregator.py]: This fails in python2 with
`TypeError: unhashable type: 'list'`
and in python3 with
`TypeError: unhashable type: 'dict_keys' `.

Comment (mstajdohar) [reply resolwe_bio/tools/expression_aggregator.py]: `set(expression.keys() for expression in expressions)`

Comment (mstajdohar) [reply resolwe_bio/tools/expression_aggregator.py]: `{key for expression in expressions for key in expression.keys()}`

Comment (mstajdohar) [reply resolwe_bio/tools/expression_aggregator.py]: This should work but it's kind-of ugly, `{key for expression in expressions for key in expression.keys()}`

File: resolwe_bio/tools/expression_aggregator.py (hunk 12 above)
Comment (mzganec) [reply resolwe_bio/tools/expression_aggregator.py]: The shebang line will be changed when the docker image with statistical python packages is ready.

File: resolwe_bio/expression_filters/sample.py [hunk 13]
```diff
@@ -21,7 +21,14 @@ def get_sample_attr(data, attr):
     if not sample_qs.exists():
         return None
 
-    return getattr(sample_qs.first(), attr, None)
+    return sample_qs.first()
+
+
+def get_sample_attr(data, attr):
+    """Get ``attr`` attribute of sample object."""
+    sample = get_sample(data)  # pylint: disable=redefined-outer-name
```
Comment (dblenkus) [inline resolwe_bio/expression_filters/sample.py]: Remove pylint disable. You can remove `sample` function and map `'sample'` to `get_sample` in `filters` dict.

File: docs/CHANGELOG.rst [hunk 14]
```diff
@@ -14,11 +14,14 @@ Added
 -----
 * Add custom Cutadapt - STAR - HT-seq workflow
 * Add ``resolwebio/rnaseq`` docker image
+* Add expression aggregator
```
Comment (dblenkus) [inline docs/CHANGELOG.rst]: Move change log entries to corresponding commits.

File: resolwe_bio/tools/expression_aggregator.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def main():
+    """Compute expression statistics."""
+    args = get_args()
+    if args.aggregator:
+        aggregator = load_json(args.aggregator, True)
+        check_aggregator(aggregator, args.source, args.expression_type, args.group_by)
+    else:
+        aggregator = None
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Put `aggregator=None` above the if and you can omit `else`.

File: resolwe_bio/tools/expression_aggregator.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        with open(fname, 'w') as f:
+            json.dump(statistics, f)
+    else:
+        with gzip.open(fname, 'wt') as f:
+            json.dump(statistics, f)
+
+
+def load_json(fname, compressed=False):
+    """Read json from file."""
+    if not compressed:
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: load_fn = gzip.open if compressed else json.load
    with open(fname, 'r') as f:
        return load_fn(f)

File: resolwe_bio/tools/expression_aggregator.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        'q3': q3,
+        'lowerwhisker': lowerwhisker,
+        'upperwhisker': upperwhisker
+    }
+
+
+def get_statistics(expressions, expression_type):
+    """Get box plot statistics for expressions of all genes and attributes."""
+    return {
+        gene: [generate_statistic(exp, gene, descriptor, expression_type) for descriptor, exp in expression.items()]
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Split this into multiple lines:

    gene: [
        generate_statistic(exp, gene, descriptor, expression_type) 
        for descriptor, exp in expression.items()
    ]

File: resolwe_bio/tools/expression_aggregator.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+def load_expressions(aggregator, expression_fns=[], sep='\t', descriptors=[]):
+    """Read expressions from files."""
+    raw_expressions = [load_expression(expression_fn, sep) for expression_fn in expression_fns]
+    if aggregator:
+        raw_expressions.extend(aggregator['raw_expressions'])
+        descriptors.extend(aggregator['descriptors'])
+    genes = {key for raw_expression in raw_expressions for key in raw_expression.keys()}
+    return raw_expressions, descriptors, {
+        gene: {
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Save the dict into a variable and then return it

File: resolwe_bio/processes/expression/expression_aggregator.yml [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        {% if e.source != exps.0.source %}
+          re-error "All expressions must have the same source."
+        {% endif %}
+        {% if e.exp_type != exps.0.exp_type %}
+          re-error "All expressions must be of the same type."
+        {% endif %}
+      {% endfor %}
+
+      expression_aggregator.py \
+        -e {% for e in exps %} {{e.exp.file}} {% endfor %} \
```
Comment (dblenkus) [inline resolwe_bio/processes/expression/expression_aggregator.yml]: Use long names for flags.

File: resolwe_bio/tools/expression_aggregator.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        'attribute': attribute,
+        'gene': gene,
+        'exp_types': [expression_type],
+        'min': min_val,
+        'max': max_val,
+        'median': median,
+        'q1': q1,
+        'q3': q3,
+        'lowerwhisker': lowerwhisker,
+        'upperwhisker': upperwhisker
```
Comment (dblenkus) [inline resolwe_bio/tools/expression_aggregator.py]: Add comma at the end of the line.

File: resolwe_bio/tools/expression_aggregator.py (hunk 16 above)
Comment (mzganec) [reply resolwe_bio/tools/expression_aggregator.py]: This fails with
```
Traceback (most recent call last):
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 161, in <module>
    main()
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 146, in main
    aggregator = load_json(args.aggregator, True)
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 117, in load_json
    return load_fn(f)
  File "/usr/lib/python2.7/gzip.py", line 34, in open
    return GzipFile(filename, mode, compresslevel)
  File "/usr/lib/python2.7/gzip.py", line 94, in __init__
    fileobj = self.myfileobj = __builtin__.open(filename, mode or 'rb')
TypeError: coercing to Unicode: need string or buffer, file found
```

**Comment by mstajdohar** [comment]
@dblenkus Please do a second review. The changes you requested have been addressed.

**Pull request merged**